### ISSUE TITLE
Assignment operator cannot be overloaded with types other than the same struct

### ISSUE THREAD

**Issue by pow2clk** [microsoft/DirectXShaderCompiler#4005]
A slight modification of the assign operator overloading test (https://github.com/microsoft/DirectXShaderCompiler/blob/master/tools/clang/test/HLSLFileCheck/hlsl/operator_overloading/operator.overloading.assign.hlsl)  reveals a limitation of its implementation:

```c++
struct Number {
    int n;

    void operator=(float x) {
        n = x;
    }
};

int main(float4 pos: SV_Position) : SV_Target {
    Number a = {pos.x};
    Number b = {pos.y};
    a = pos.x;
    return a.n;
}
```
error: cannot implicitly convert from 'float' to 'Number'
    a = pos.x;

This is permitted in C++ and would be useful here as well.

**Comment by jaebaek** [comment]
I am taking a look at this issue. It seems to be related to both #4004 and #4066

**Comment by jaebaek** [comment]
This was fixed by #4063
